### ISSUE TITLE
prepareForAsyncCopy crashes if a program compiled for CUDA runs without a GPU

### ISSUE THREAD

**Issue by fwyzard** [alpaka-group/alpaka#1588]
I'm not sure if this can be addressed inside alpaka, but we have run into a problem with this scenario.

If a program is built with support for CUDA (setting `ALPAKA_ACC_GPU_CUDA_ENABLED`), but runs on a machine without access to a GPU (_e.g._ if `CUDA_VISIBLE_DEVICES` is set to an empty list at runtime), a call to `alpaka::prepareForAsyncCopy(buffer)` will result in an exception:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  /data/user/fwyzard/pixeltrack-standalone/external/alpaka/include/alpaka/mem/buf/BufCpu.hpp(334) 'ALPAKA_API_PREFIX(HostRegister)( const_cast<void*>(reinterpret_cast<void const*>(getPtrNative(buf))), extent::getExtentProduct(buf) * sizeof(Elem<BufCpu<TElem, TDim, TIdx>>), ALPAKA_API_PREFIX(HostRegisterDefault))' A previous API call (not this one) set the error  : 'cudaErrorNoDevice': 'no CUDA-capable device is detected'!
Aborted (core dumped)
```

The reason is clear (the call to `cudaHostRegister` fails because the CUDA runtime does not initialise if there is no GPU).

What is less clear if this is the desired behaviour, or if in this case `prepareForAsyncCopy` should simply do nothing.

**Comment by bussmann** [comment]
Could you elaborate on the use case scenario for which this is happening? Why do you need a CUDA backend when there is no GPU present?

**Comment by fwyzard** [comment]
Sure: the use case is for building an application once, with support for multiple backends (gpus, cpus), and being able to deploy it on different kind of machines, with or without gpus.

**Comment by tonydp03** [comment]
We discussed today in VC and seems that this issue and #1613 could be solved changing some Makefile logic for compiling alpaka applications. @psychocoderHPC

**Comment by j-stephan** [comment]
`prepareForAsyncCopy` needs to be changed internally, too. I guess that it should be enough to check for the number of CUDA devices in the system and only act if that number is greater than 0.

**Comment by tonydp03** [comment]
That's also what I was thinking. I didn't see any other solution than checking the list of available CUDA devices.

**Comment by fwyzard** [comment]
I expect that the list of available devices should not change during a program's execution, so would it make sense to discover this information once and store it in a central place ?

**Comment by bussmann** [comment]
I would not make this assumption. Combinig Alpaka with virtualization would invalidate this assumption. Think of dynamic resource allocation by a virtualized HPC system with several demanding workloads running in parallel. You might have 256 GPUs to start with and then end up with 192 mid run. Currently, this is not a scenario seen in the wild very often, if at all, but with more demanding work and limited resources at high data throughput this may happen.

**Comment by psychocoderHPC** [comment]
> I expect that the list of available devices should not change during a program's execution, so would it make sense to discover this information once and store it in a central place ?

IMO this is application dependent and should not be part of alpaka.

**Comment by psychocoderHPC** [comment]
> We discussed today in VC and seems that this issue and #1613 could be solved changing some Makefile logic for compiling alpaka applications. @psychocoderHPC

Currently alpaka's accelerators are enabled via CMake variables, we should provide alpaka targets for each accelerator. This would open the user to deside which shared library, static library or binary should contain which accelerator.
This should also solve the `prepareForAsyncCopy`.

**Comment by fwyzard** [comment]
> IMO this is application dependent and should not be part of alpaka.

I would rather say that it is backend dependent:
  - as far as I know TBB does not allow to change the number of worker threads once a pool has been created
  - the SYCL documentation says explicitly that 
    > The execution environment for a SYCL application has a fixed number of devices which does not vary as the application executes.
  - I'd be surprised if CUDA or ROCm can discover new devices at runtime - but I've never tried...

Anyway, what I meant is that this change

> `prepareForAsyncCopy` needs to be changed internally, too. I guess that it should be enough to check for the number of CUDA devices in the system and only act if that number is greater than 0.

risks making `prepareForAsyncCopy` extremely expensive to call, if it checks for new devices every time.

**Comment by psychocoderHPC** [comment]
> > I expect that the list of available devices should not change during a program's execution, so would it make sense to discover this information once and store it in a central place ?
> 
> IMO this is application dependent and should not be part of alpaka.

@fwyzard I misinterpreted your comment. I thought you like to store the user-chosen device somewhere, instead, you mean you would like to query all possible devices only once and collect all information and if the user is executing the query again the information should be provided from the cached information.

I agree we should think about it.

**Comment by fwyzard** [comment]
Specifically for the `prepareForAsyncCopy` case, an other option could be to change to function signature to take as a second parameter the `Queue` or `Device` that the memory should be pinned _for_ :
```c++
auto buffer = alpaka::allocBuf<T, Idx>(host, ,..);
alpaka::prepareForAsyncCopy(buffer, device);
```

I'm the first to say that it may be annoying - but it may also be (a) necessary (extension) for #1613 .

**Comment by fwyzard** [comment]
In the context of #1613, an other alternative could be
  - if `ALPAKA_ACC_GPU_CUDA_ENABLED` is defined and the system has any NVIDIA GPUs, pin the memory with `cudaHostRegister` 
  - if `ALPAKA_ACC_GPU_HIP_ENABLED` is enabled and the system has any AMD GPUs, pin the memory with `hipHostRegister`
  - if both are true, pin the memory twice, with both calls

Strange as it is, pinning both at the same time it did work on my test system with a GeForce GTX 1080 Ti and a Radeon Pro WX 9100, and I could access the memory directly from both GPUs.

Though, the results may of course differ with different GPU models, and I did _not_ try to perform atomic operations on the pinned memory from both GPUs at the same time...

**Comment by fwyzard** [comment]
Fixed by #1790, removing the support for pinning/unpinning existing buffers.
Use `allocMappedBuf()` instead.
